Public/Src/Engine/Scheduler/Distribution/Worker.cs

```diff
@@ -421,27 +421,54 @@ internal bool TryAcquire(RunnablePip runnablePip, out WorkerResource? limitingRe
+            var acquiredSlots = m_acquiredProcessSlots;
+            var totalSlotsNeeded = acquiredSlots + processRunnablePip.Process.Weight;
+
+            // If a process has a weight higher than the total number of process slots, still allow it to run as long as there are no other
+            // processes running (the number of acquired slots is 0)
+            bool processSlotsAcquired = false;
+            if (acquiredSlots == 0
+                && Interlocked.CompareExchange(ref m_acquiredProcessSlots, processRunnablePip.Process.Weight, 0) == 0)
             {
-                limitingResource = WorkerResource.AvailableProcessSlots;
-                return false;
+                processSlotsAcquired = true;
+            }
+            else if (totalSlotsNeeded <= (EffectiveTotalProcessSlots * loadFactor)
+                && Interlocked.CompareExchange(ref m_acquiredProcessSlots, totalSlotsNeeded, acquiredSlots) == acquiredSlots)
+            {
+                processSlotsAcquired = true;
             }
 
-            var processRunnablePip = runnablePip as ProcessRunnablePip;
+            Contract.Assert(processRunnablePip != null);
             StringId limitingResourceName = StringId.Invalid;
-            if (processRunnablePip != null &&
-                processRunnablePip.TryAcquireResources(m_workerSemaphores, GetAdditionalResourceInfo(processRunnablePip), out limitingResourceName))
+            if (processSlotsAcquired && processRunnablePip.TryAcquireResources(m_workerSemaphores, GetAdditionalResourceInfo(processRunnablePip), out limitingResourceName))
             {
-                Interlocked.Increment(ref m_acquiredProcessSlots);
                 OnWorkerResourcesChanged(WorkerResource.AvailableProcessSlots, increased: false);
                 runnablePip.AcquiredResourceWorker = this;
                 limitingResource = null;
                 return true;
             }
 
-            limitingResource = limitingResourceName == m_ramSemaphoreNameId
-                ? WorkerResource.AvailableMemoryMb
-                : WorkerResource.CreateSemaphoreResource(limitingResourceName.ToString(runnablePip.Environment.Context.StringTable));
+            // Failure cases
+            if (processSlotsAcquired)
+            {
+                // Release the acquired slots
+                Interlocked.Add(ref m_acquiredProcessSlots, -processRunnablePip.Process.Weight);
+
+                if (limitingResourceName == m_ramSemaphoreNameId)
+                {
+                    limitingResource = WorkerResource.AvailableMemoryMb;
+                }
+                else
+                {
+                    limitingResource = WorkerResource.CreateSemaphoreResource(limitingResourceName.ToString(runnablePip.Environment.Context.StringTable));
+                }
+            }
+            else
+            {
+                limitingResource = WorkerResource.AvailableProcessSlots;

```

Comment:
>limitingResource = WorkerResource.AvailableProcessSlots; [](start = 16, length = 56)

I'm thinking it would be good to get the number of process slots pips use into one of their log messages to help debug the "why aren't more pips running?" case. Sticking it in the pip description may be overkill. What do you think about putting it in the ProcessPipCacheMiss message?

Reply:
Do you mean this message? 




---
In reply to: [271090384](https://github.com/Microsoft/BuildXL/pull/67#discussion_r271090384) [](ancestors = 271090384)

Reply:
yes. i see you added a separate message. That's probably fine too as long as this doesn't get widely used. We may want to revisit consolidating it into an existing message to save space if it is used at lot.

---
In reply to: [271094601](https://github.com/Microsoft/BuildXL/pull/67#discussion_r271094601) [](ancestors = 271094601,271090384)